executor/set_config.go

```diff
@@ -0,0 +1,173 @@
+		return core.ErrSpecificAccessDenied.GenWithStackByArgs("SET CONFIG")
+	}
+
+	if s.p.Type != "" {
+		s.p.Type = strings.ToLower(s.p.Type)
+		if s.p.Type != "tikv" && s.p.Type != "tidb" && s.p.Type != "pd" {
+			return errors.Errorf("unknown type %v", s.p.Type)
+		}
+		if s.p.Type == "tidb" {
+			return errors.Errorf("TiDB doesn't support to change configs online, please use SQL variables")
+		}
+	}
+	if s.p.Instance != "" {
+		s.p.Instance = strings.ToLower(s.p.Instance)
+		if !isValidInstance(s.p.Instance) {
+			return errors.Errorf("invalid instance %v", s.p.Instance)
+		}
+	}
+	s.p.Name = strings.ToLower(s.p.Name)
+
+	val, isNull, err := s.p.Value.EvalString(s.ctx, chunk.Row{})
+	if err != nil {
+		return err
+	}
+	if isNull {
+		return errors.Errorf("can't set config to null")
+	}
+	s.v = val
+	return nil
+}
+
+// TestSetConfigServerInfoKey is used as the key to store 'TestSetConfigServerInfoFunc' in the context.
+var TestSetConfigServerInfoKey stringutil.StringerStr = "TestSetConfigServerInfoKey"
+
+// TestSetConfigHTTPHandlerKey is used as the key to store 'TestSetConfigDoRequestFunc' in the context.
+var TestSetConfigHTTPHandlerKey stringutil.StringerStr = "TestSetConfigHTTPHandlerKey"
+
+// Next implements the Executor Next interface.
+func (s *SetConfigExec) Next(ctx context.Context, req *chunk.Chunk) error {
+	req.Reset()
+	getServerFunc := infoschema.GetClusterServerInfo
+	if v := s.ctx.Value(TestSetConfigServerInfoKey); v != nil {
+		getServerFunc = v.(func(sessionctx.Context) ([]infoschema.ServerInfo, error))
+	}
+
+	serversInfo, err := getServerFunc(s.ctx)
+	if err != nil {
+		return err
+	}
+	nodeTypes := set.NewStringSet()
+	nodeAddrs := set.NewStringSet()
+	if s.p.Type != "" {
+		nodeTypes.Insert(s.p.Type)
+	}
+	if s.p.Instance != "" {
+		nodeAddrs.Insert(s.p.Instance)
+	}
+	serversInfo = filterClusterServerInfo(serversInfo, nodeTypes, nodeAddrs)
+
+	for _, serverInfo := range serversInfo {
+		var url string
+		switch serverInfo.ServerType {
+		case "pd":
+			url = fmt.Sprintf("%s://%s%s", util.InternalHTTPSchema(), serverInfo.StatusAddr, pdapi.Config)
+		case "tikv":
+			url = fmt.Sprintf("%s://%s/config", util.InternalHTTPSchema(), serverInfo.StatusAddr)
+		default:
+			continue
+		}
+		if err := s.doRequest(url); err != nil {
+			s.ctx.GetSessionVars().StmtCtx.AppendWarning(err)
+		}
+	}
+	return nil
+}
+
+func (s *SetConfigExec) doRequest(url string) (retErr error) {
+	body := bytes.NewBufferString(fmt.Sprintf("{'%s':'%s'}", s.p.Name, s.v))
+	req, err := http.NewRequest(http.MethodPost, url, body)
+	if err != nil {
+		return err
+	}
+	var httpHandler func(req *http.Request) (*http.Response, error)
+	if v := s.ctx.Value(TestSetConfigHTTPHandlerKey); v != nil {
+		httpHandler = v.(func(*http.Request) (*http.Response, error))
+	} else {
+		httpHandler = util.InternalHTTPClient().Do
+	}
+	resp, err := httpHandler(req)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if resp.Body != nil {
```

Comment:
> If the returned error is nil, the Response will contain a non-nil Body which the user is expected to close. If the Body is not both read to EOF and closed, the Client's underlying RoundTripper (typically Transport) may not be able to re-use a persistent TCP connection to the server for a subsequent "keep-alive" request.

Not necessary to check `resp.Body` here.

Reply:
Fixed. PTAL.